docs/design/2020-06-27-non-recursive-common-table-expression.md

```diff
@@ -0,0 +1,188 @@
+* Optimizer hints specifies [NO_MERGE](https://dev.mysql.com/doc/refman/8.0/en/optimizer-hints.html#optimizer-hints-table-level), or [optimizer_switch](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_optimizer_switch)'s `derived_merge` is __OFF__.
+* CTE is not mergeable because having the following structure [[16]](https://github.com/mysql/mysql-server/blob/f8cdce86448a211511e8a039c62580ae16cb96f5/sql/sql_lex.cc#L3727):
+  *  Union (implementation restriction) _(TODO: why?)_
+  *  Aggregation/Having/Distinct (A general aggregation cannot be merged with a non-aggregated one)
+  *  Limit
+  *  Window _(TODO: why?)_
+* Heuristic does not suggest merge [[16]](https://github.com/mysql/mysql-server/blob/f8cdce86448a211511e8a039c62580ae16cb96f5/sql/sql_lex.cc#L3751):
+  * CTE is not constant (having user-defined variable assignment).
+  * Contains subqueries in the SELECT list that depend on columns from itself, because MySQL assumes that user wants subquery to be materialized.
+
+#### MariaDB
+MariaDB using `Materialization` as basic algorithm. Similar to MySQL, to optimize performance, MariaDB will always try to use `Merge` strategy, and fall back to materialization due to side-effect [[18]](https://github.com/MariaDB/server/blob/5b8ab1934a10966336e66751bc13fc66923b02f6/sql/table.cc#L9224).
+
+Besides, MariaDB supports "Condition Pushdown" [[5]](http://ceur-ws.org/Vol-1864/paper_6.pdf) and "Lateral Derived Optimization" [[18]](https://www.slideshare.net/SergeyPetrunya/mariadb-103-optimizer-where-does-it-stand)[[19]](https://mariadb.com/kb/en/lateral-derived-optimization/)
+
+#### Greenplum
+Greenplum also using `Merge` and `Materialization`, and utilize rules and cost-bases algorithm to optimize execution plan. Moreover, Greenplum introduces a producer-consumer model to implement `Materialization` [[4]](http://www.vldb.org/pvldb/vol8/p1704-elhelw.pdf).
+
+#### Postgresql
+Postgresql have always used `Materialization` strategy, and introduced `Merge` in a recent commit [[9]](https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=608b167f9f9c4553c35bb1ec0eab9ddae643989b)[[10]](https://info.crunchydata.com/blog/with-queries-present-future-common-table-expressions).
+
+#### CockroachDB
+CockroachDB uses `Materialization` strategy by default, but chooses `Merge` strategy when 1) see `NO MATERIALIZED` hint, or 2) the CTE has no side-effect and is referenced only once [[11]](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/norm/with_funcs.go#L18)[[12]](https://github.com/cockroachdb/cockroach/issues/45863).
+
+
+## Implementation
+
+1. Stage 1: `Merge` should be not difficult. A rewrite rule is enough, similar to rewriting subquery [[2]](https://pingcap.com/blog/2016-12-07-Subquery-Optimization-in-TiDB/).
+2. Stage 2: Implement `Materialization`, by storing chunks temporarily in memory or disk.
+3. Stage 3: `Predicate Push-down` for `Materialization`, and estimate cost for both strategies.
```

Comment:
What does this mean? When it becomes a physical tmp table, the predicate cannot be pushed into it.